### ISSUE TITLE
PGX v5 Pool poor performance on Windows

### ISSUE THREAD

**Issue by codercms** [jackc/pgx#1523]
**Describe the bug**
When using the pgx pool on Windows, the query time is too long (e.g. `SELECT 1` query can take up to 120ms)

**To Reproduce**
Run code example below

<details>
<summary>Code (click me)</summary>

```go
package main

import (
	"context"
	"fmt"
	pgxv4 "github.com/jackc/pgx/v4"
	pgxv4pool "github.com/jackc/pgx/v4/pgxpool"
	pgxv5 "github.com/jackc/pgx/v5"
	pgxv5pool "github.com/jackc/pgx/v5/pgxpool"
	"os"
	"time"
)

func main() {
	connStr := os.Getenv("DATABASE_URL")

	fmt.Printf("pgx v5 conn:\n")
	testPgxV5Conn(connStr)

	fmt.Printf("\npgx v5 pool:\n")
	testPgxV5Pool(connStr)

	fmt.Printf("\npgx v4 conn:\n")
	testPgxV4Conn(connStr)

	fmt.Printf("\npgx v4 pool:\n")
	testPgxV4Pool(connStr)
}

func testPgxV5Pool(connStr string) {
	cfg, err := pgxv5pool.ParseConfig(connStr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Unable to parse pool config: %v\n", err)
		os.Exit(1)
	}

	cfg.MaxConns = 1

	conn, err := pgxv5pool.NewWithConfig(context.Background(), cfg)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Unable to connect to database: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close()

	if err := conn.Ping(context.Background()); err != nil {
		fmt.Fprintf(os.Stderr, "pool ping failed: %v\n", err)
		os.Exit(1)
	}

	makeAttemptPool(conn, 0)

	time.Sleep(time.Second)

	makeAttemptPool(conn, 1)

	time.Sleep(time.Second)

	makeAttemptPool(conn, 2)
}

func testPgxV5Conn(connStr string) {
	conn, err := pgxv5.Connect(context.Background(), connStr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Unable to connect to database: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close(context.Background())

	makeAttempt(conn, 0)

	time.Sleep(time.Second)

	makeAttempt(conn, 1)

	time.Sleep(time.Second)

	makeAttempt(conn, 2)
}

func testPgxV4Pool(connStr string) {
	cfg, err := pgxv4pool.ParseConfig(connStr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Unable to parse pool config: %v\n", err)
		os.Exit(1)
	}

	cfg.MaxConns = 1

	conn, err := pgxv4pool.ConnectConfig(context.Background(), cfg)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Unable to connect to database: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close()

	if err := conn.Ping(context.Background()); err != nil {
		fmt.Fprintf(os.Stderr, "pool ping failed: %v\n", err)
		os.Exit(1)
	}

	makeAttemptPoolV4(conn, 0)

	time.Sleep(time.Second)

	makeAttemptPoolV4(conn, 1)

	time.Sleep(time.Second)

	makeAttemptPoolV4(conn, 2)
}

func testPgxV4Conn(connStr string) {
	conn, err := pgxv4.Connect(context.Background(), connStr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Unable to connect to database: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close(context.Background())

	makeAttemptV4(conn, 0)

	time.Sleep(time.Second)

	makeAttemptV4(conn, 1)

	time.Sleep(time.Second)

	makeAttemptV4(conn, 2)
}

func makeAttempt(conn *pgxv5.Conn, attempt int) {
	start := time.Now()

	var id int64

	err := conn.QueryRow(context.Background(), "select 1 AS id").Scan(&id)
	if err != nil {
		fmt.Fprintf(os.Stderr, "QueryRow failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("v5 conn time took #%d = %s\n", attempt, time.Since(start))
}

func makeAttemptPool(conn *pgxv5pool.Pool, attempt int) {
	start := time.Now()

	var id int64

	err := conn.QueryRow(context.Background(), "select 1 AS id").Scan(&id)
	if err != nil {
		fmt.Fprintf(os.Stderr, "QueryRow failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("v5 pool time took #%d = %s\n", attempt, time.Since(start))
}

func makeAttemptV4(conn *pgxv4.Conn, attempt int) {
	start := time.Now()

	var id int64

	err := conn.QueryRow(context.Background(), "select 1 AS id").Scan(&id)
	if err != nil {
		fmt.Fprintf(os.Stderr, "QueryRow failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("v4 conn time took #%d = %s\n", attempt, time.Since(start))
}

func makeAttemptPoolV4(conn *pgxv4pool.Pool, attempt int) {
	start := time.Now()

	var id int64

	err := conn.QueryRow(context.Background(), "select 1 AS id").Scan(&id)
	if err != nil {
		fmt.Fprintf(os.Stderr, "QueryRow failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("v4 pool time took #%d = %s\n", attempt, time.Since(start))
}
```
</details>

**Expected behavior**
pgx pool shouldn't add so much delay

**Actual behavior**
pgx pool adds extra delay
```
pgx v5 conn:
v5 conn time took #0 = 1.2867ms
v5 conn time took #1 = 0s
v5 conn time took #2 = 86.7µs

pgx v5 pool:
v5 pool time took #0 = 505.3µs
v5 pool time took #1 = 111.0746ms
v5 pool time took #2 = 109.5005ms

pgx v4 conn:
v4 conn time took #0 = 701.8µs
v4 conn time took #1 = 639.5µs
v4 conn time took #2 = 130.9µs

pgx v4 pool:
v4 pool time took #0 = 744.9µs
v4 pool time took #1 = 0s
v4 pool time took #2 = 135.3µs
```

Here is timings for Ubuntu (go 1.20):
```
pgx v5 conn:
v5 conn time took #0 = 440.4µs
v5 conn time took #1 = 292.9µs
v5 conn time took #2 = 395.5µs

pgx v5 pool:
v5 pool time took #0 = 503.7µs
v5 pool time took #1 = 338.3µs
v5 pool time took #2 = 346µs

pgx v4 conn:
v4 conn time took #0 = 7.0436ms
v4 conn time took #1 = 282.7µs
v4 conn time took #2 = 271.4µs

pgx v4 pool:
v4 pool time took #0 = 424.5µs
v4 pool time took #1 = 383.5µs
v4 pool time took #2 = 284.3µs
```

**Version**
 - Go: `go version go1.19 windows/amd64` and also tested on `1.20` (win, x64)
 - PostgreSQL: `PostgreSQL 14.5, compiled by Visual C++ build 1914, 64-bit`
 - pgx: `v5.3.0`

**Comment by jackc** [comment]
This is almost certainly caused by the use of fake non-blocking IO on Windows (see #1482).

Your `time.Sleep(time.Second)` makes it worst case though. See https://github.com/jackc/pgx/blob/38e09bda4c68fc87967b50455640e308bda6a959/pgxpool/pool.go#L498-L504. The pool tries to check if a connection is alive when it is checked out. If it has been idle for 1+ second it tries a non-blocking read. You can see this in your first pool request on Windows being fast and the subsequent ones are slower.

It also may improve with further use as it adapts to find the shortest time to wait for a fake non-blocking read to fail.

**Comment by codercms** [comment]
@jackc I see, well, I use pgx for API and when API is not under load response times are really bad, because of fake non-blocking I/O ~100-120ms latency.

I made research of Winsock library and implemented non-blocking read function (based on linux non block function):
```go
// new imports needed:
import "golang.org/x/sys/windows"
import "unsafe"
import "io"

var dll = syscall.NewLazyDLL("ws2_32.dll")

//int ioctlsocket(
//	[in]      SOCKET s,
//	[in]      long   cmd,
//	[in, out] u_long *argp
//);
var ioctlsocket = dll.NewProc("ioctlsocket")

const FIONBIO = 0x8004667e

// patched fakeNonblockingRead in nbconn package
func (c *NetConn) fakeNonblockingRead(b []byte) (n int, err error) {
	if c.nonblockReadFunc == nil {
		c.nonblockReadFunc = func(fd uintptr) (done bool) {
			var nonBlock int = 1

			//u_long mode = 1;  // 1 to enable non-blocking socket
			//ioctlsocket(sock, FIONBIO, &mode);
			// Make sock non-blocking
			ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&nonBlock)))

			var tmpBuf [1]byte
			buf := &windows.WSABuf{
				Len: 1,
				Buf: &tmpBuf[0],
			}

			var recvd uint32
			var flags uint32
			croutine := byte(0)

			c.nonblockReadErr = windows.WSARecv(windows.Handle(fd), buf, 1, &recvd, &flags, nil, &croutine)
			c.nonblockReadN = int(recvd)

			// Make sock blocking again
			nonBlock = 0
			ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&nonBlock)))

			//c.nonblockReadN, c.nonblockReadErr = syscall.Read(syscall.Handle(fd), c.nonblockReadBuf)
			return true
		}
	}
	c.nonblockReadBuf = b
	c.nonblockReadN = 0
	c.nonblockReadErr = nil

	err = c.rawConn.Read(c.nonblockReadFunc)
	n = c.nonblockReadN
	c.nonblockReadBuf = nil // ensure that no reference to b is kept.
	if err == nil && c.nonblockReadErr != nil {
		if errors.Is(c.nonblockReadErr, windows.WSAEWOULDBLOCK) {
			err = ErrWouldBlock
		} else {
			err = c.nonblockReadErr
		}
	}
	if err != nil {
		// n may be -1 when an error occurs.
		if n < 0 {
			n = 0
		}

		return n, err
	}

	// syscall read did not return an error and 0 bytes were read means EOF.
	if n == 0 {
		return 0, io.EOF
	}

	return n, nil
}
```

Maybe this could be helpful to you, to fix non-blocking I/O for windows.  
This function seems to work, but I haven't tested it much, but response time decreased drastically.  
And I'm not sure if it will work in x86 windows, because I don't have deep understanding how's GoLang FFI is working.

But actually I don't understand where's going read bytes in linux non-blocking I/O implementation

**Comment by codercms** [comment]
BTW: I'm not even sure if such approach will work on Windows, I mean - check connection is alive by trying to read from socket.
Because Windows has completely different network stack, and I don't know what will happen if there are some buffered bytes in socket, will they be read from socket, or will it return an error

**Comment by jackc** [comment]
Try branch [windows-non-blocking-io](https://github.com/jackc/pgx/tree/windows-non-blocking-io). I did not explicitly set the socket to non-blocking as I think that the Go runtime should have the sockets in non-blocking mode already. But I'm really out of my element on Windows so who knows.

**Comment by codercms** [comment]
@jackc socket should be really explicitly set to be non-blocking, because on Windows default socket mode is blocking, so when there is no bytes available in socket and when no recv timeout set  `WSARecv` would block forever until some bytes arrive. (of course I've tested this behavior, code from branch `windows-non-blocking-io` just hangs forever).

And I believe this check `if errors.Is(c.nonblockReadErr, syscall.EWOULDBLOCK` should be written as `errors.Is(c.nonblockReadErr, windows.WSAEWOULDBLOCK)`

`syscall.EWOULDBLOCK` resolved value is `536871039`
`windows.WSAEWOULDBLOCK` resolved value is `10035`

Because WSA functions has different error codes, e.g.: WSAEWOULDBLOCK, 10035

* Here is WSARecv documentation itself - https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsarecv
* Here is WSA error codes list - https://learn.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2
* And here is about errors handling in WSA: https://learn.microsoft.com/en-us/windows/win32/winsock/return-values-on-function-failure-2

**Comment by codercms** [comment]
@jackc I have locally patched `nbconn_real_non_block_windows.go` file to:

<details>
<summary>Click Me</summary>

```go
//go:build windows

package nbconn

import (
	"errors"
	"fmt"
	"io"
	"syscall"
	"unsafe"
	"golang.org/x/sys/windows"
)

var dll = syscall.MustLoadDLL("ws2_32.dll")

//int ioctlsocket(
//	[in]      SOCKET s,
//	[in]      long   cmd,
//	[in, out] u_long *argp
//);
var ioctlsocket = dll.MustFindProc("ioctlsocket")

const FIONBIO = 0x8004667e

// realNonblockingWrite does a non-blocking write. readFlushLock must already be held.
func (c *NetConn) realNonblockingWrite(b []byte) (n int, err error) {
	if c.nonblockWriteFunc == nil {
		c.nonblockWriteFunc = func(fd uintptr) (done bool) {
			var nonBlock int = 1

			// Make sock non-blocking
			ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&nonBlock)))

			var written uint32
			var buf syscall.WSABuf
			buf.Buf = &c.nonblockWriteBuf[0]
			buf.Len = uint32(len(c.nonblockWriteBuf))
			c.nonblockWriteErr = syscall.WSASend(syscall.Handle(fd), &buf, 1, &written, 0, nil, nil)
			c.nonblockWriteN = int(written)

			// Make sock blocking again
			nonBlock = 0
			ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&nonBlock)))

			return true
		}
	}
	c.nonblockWriteBuf = b
	c.nonblockWriteN = 0
	c.nonblockWriteErr = nil

	err = c.rawConn.Write(c.nonblockWriteFunc)
	n = c.nonblockWriteN
	c.nonblockWriteBuf = nil // ensure that no reference to b is kept.
	if err == nil && c.nonblockWriteErr != nil {
		if errors.Is(c.nonblockWriteErr, windows.WSAEWOULDBLOCK){
			fmt.Println("WSA WOULD BLOCK catched")
			err = ErrWouldBlock
		} else {
			fmt.Printf("WSA WOULD BLOCK not catched, :%v", err)
			err = c.nonblockWriteErr
		}
	}
	if err != nil {
		// n may be -1 when an error occurs.
		if n < 0 {
			n = 0
		}

		return n, err
	}

	return n, nil
}

func (c *NetConn) realNonblockingRead(b []byte) (n int, err error) {
	if c.nonblockReadFunc == nil {
		c.nonblockReadFunc = func(fd uintptr) (done bool) {
			var nonBlock int = 1

			// Make sock non-blocking
			ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&nonBlock)))

			var read uint32
			var flags uint32
			var buf syscall.WSABuf
			buf.Buf = &c.nonblockReadBuf[0]
			buf.Len = uint32(len(c.nonblockReadBuf))
			c.nonblockReadErr = syscall.WSARecv(syscall.Handle(fd), &buf, 1, &read, &flags, nil, nil)
			c.nonblockReadN = int(read)

			// Make sock blocking again
			nonBlock = 0
			ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&nonBlock)))

			return true
		}
	}
	c.nonblockReadBuf = b
	c.nonblockReadN = 0
	c.nonblockReadErr = nil

	err = c.rawConn.Read(c.nonblockReadFunc)
	n = c.nonblockReadN
	c.nonblockReadBuf = nil // ensure that no reference to b is kept.
	if err == nil && c.nonblockReadErr != nil {
		if errors.Is(c.nonblockReadErr, windows.WSAEWOULDBLOCK) {
			fmt.Println("WSA WOULD BLOCK catched")
			err = ErrWouldBlock
		} else {
			fmt.Printf("WSA WOULD BLOCK not catched, :%v", err)
			err = c.nonblockReadErr
		}
	}
	if err != nil {
		// n may be -1 when an error occurs.
		if n < 0 {
			n = 0
		}

		return n, err
	}

	// syscall read did not return an error and 0 bytes were read means EOF.
	if n == 0 {
		return 0, io.EOF
	}

	return n, nil
}

```
</details>

It seems to work and I see in console output `WSA WOULD BLOCK catched`
